driver-core/src/test/unit/com/mongodb/internal/logging/LogMessageTest.java

```diff
@@ -0,0 +1,163 @@
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+
+class UnstructuredLogMessageTest {
+
+    @ParameterizedTest
+    @MethodSource("provideExpectedWithEntries")
+    void shouldInterpolateMessage(String format, String expectedMessage, List<LogMessage.Entry> entries) {
+        LogMessage.UnstructuredLogMessage unstructuredLogMessage = new LogMessage(LogMessage.Component.COMMAND, LogMessage.Level.DEBUG,
+                "Connection id", new ClusterId(), entries, format).toUnstructuredLogMessage();
+
+        String actualMessage = unstructuredLogMessage.interpolate();
+        Assertions.assertEquals(expectedMessage, actualMessage);
+    }
+
+    private static Stream<Arguments> provideExpectedWithEntries() {
+
+       String format =  "Command \"{}\" succeeded in {} ms using a connection with driver-generated ID {}" +
+                "[ and server-generated ID {}] to {}:{}[ with service ID {}]. The requestID is {} and the " +
+                "operation ID is {}. Command reply: {}]";
```

Comment:
Good catch! As we discussed, the current algorithm ignores the special symbol ']' since it lacks a match, which allowed the tests to pass. While it is essential to consider covering edge cases and flagging errors, in our specific case, we concluded that having tests to verify the correctness of the log output is sufficient for ensuring the expected functionality without unnecessarily complicating the interpolation process. However, we can keep it in mind for future improvements if needed

Reply:
Good catch! As we discussed, the current algorithm ignores the special symbol ']' since it lacks a match, which allowed the tests to pass. While it is essential to consider covering edge cases and flagging errors, in our specific case, we concluded that having tests to verify the correctness of the log output is sufficient for ensuring the expected functionality without unnecessarily complicating the interpolation process. However, we can keep it in mind for future improvements if needed